src/Common/CaresPTRResolver.cpp

```diff
@@ -45,7 +50,12 @@ namespace DB
     CaresPTRResolver::~CaresPTRResolver()
     {
         ares_destroy(channel);
-        ares_library_cleanup();
+        /*
+         * Library initialization is currently done only once in the constructor. Multiple instances of CaresPTRResolver
+         * will be used in the lifetime of ClickHouse, thus it's problematic to have de-init here.
+         * In a practical view, it makes little to no sense to de-init a DNS library since DNS requests will happen
+         * until the end of the program. Hence, ares_library_cleanup() will not be called.
+         * */
```

Comment:
Does `ares_library_cleanup` free some memory? It may cause test failures with LeakSanitizer

Reply:
As of now, it doesn't. It only re-sets a couple of global variables, see implementation: https://github.com/c-ares/c-ares/blob/main/src/lib/ares_library_init.c#L85. Plus, `ares_library_init` doesn't seem allocate anything, see implementation: https://github.com/c-ares/c-ares/blob/main/src/lib/ares_library_init.c#L54.

In general, it's not a good thing to rely on these implementation details, but it looks better than alternative approaches like having some sort of lock to call `ares_library_init` & `ares_library_cleanup` multiple times in a thread safe manner.